Zend/zend_vm_def.h

```diff
@@ -501,7 +501,11 @@ ZEND_VM_HELPER(zend_is_equal_helper, ANY, ANY, zval *op_1, zval *op_2)
 	if (UNEXPECTED(Z_TYPE_INFO_P(op_2) == IS_UNDEF)) {
 		op_2 = ZVAL_UNDEFINED_OP2();
 	}
-	ret = zend_compare(op_1, op_2);
+	if (Z_TYPE_P(op_1) == IS_OBJECT || Z_TYPE_P(op_2) == IS_OBJECT) {
+		ret = zend_equals_object(op_1, op_2, ZEND_IS_EQUAL);
```

Comment:
I believe you need a separate zend_equals that is pushed through the entire hierarchy. Otherwise objects nested in arrays will not use the correct handler, for example.

Reply:
Ah, similar to `zend_compare` that handles all types and type combinations? I hadn't considered the case of objects in arrays.

Reply:
> Ah, similar to zend_compare that handles all types and type combinations?

Yeah, exactly.